src/solvers/sat/satcheck_ipasir.cpp

```diff
@@ -0,0 +1,200 @@
+
+  const int val=ipasir_val(solver, a.var_no());
+
+  if(val>0)
+    result=tvt(true);
+  else if(val<0)
+    result=tvt(false);
+  else
+    return tvt::unknown();
+
+  if(a.sign())
+    result=!result;
+
+  return result;
+}
+
+const std::string satcheck_ipasirt::solver_text()
+{
+  return std::string(ipasir_signature());
+}
+
+void satcheck_ipasirt::lcnf(const bvt &bv)
+{
+  forall_literals(it, bv)
+  {
+    if(it->is_true())
+      return;
+    else if(!it->is_false())
+      INVARIANT(it->var_no()<(unsigned)no_variables(),
+             "reject out of bound variables");
+  }
+
+
+  forall_literals(it, bv)
+  {
+    if(!it->is_false())
+    {
+      // add literal wit correct sign
+      ipasir_add(solver, it->sign()?-it->var_no():it->var_no());
+    }
+  }
+  ipasir_add(solver, 0); // terminate clause
+
+  clause_counter++;
+}
+
+propt::resultt satcheck_ipasirt::prop_solve()
+{
+  INVARIANT(status!=statust::ERROR, "there cannot be an error");
+
+  {
+    messaget::status() <<
+      (no_variables()-1) << " variables, " <<
+      clause_counter << " clauses" << eom;
+  }
+
+  // use the internal representation, as ipasir does not support reporting the
+  // status
+  if(status==statust::UNSAT)
+  {
+    messaget::status() <<
+      "SAT checker inconsistent: instance is UNSATISFIABLE" << eom;
+  }
+  else
+  {
+    // if assumptions contains false, we need this to be UNSAT
+    bool has_false=false;
+
+    forall_literals(it, assumptions)
+      if(it->is_false())
+        has_false=true;
+
+    if(has_false)
+    {
+      messaget::status() <<
+        "got FALSE as assumption: instance is UNSATISFIABLE" << eom;
+    }
+    else
+    {
+      forall_literals(it, assumptions)
+        if(!it->is_false())
+          ipasir_assume(solver, it->sign()?-it->var_no():it->var_no());
+
+      if(10==ipasir_solve(solver))
+      {
+        messaget::status() <<
+          "SAT checker: instance is SATISFIABLE" << eom;
+        status=statust::SAT;
+        return resultt::P_SATISFIABLE;
+      }
+      else
+      {
+        messaget::status() <<
+          "SAT checker: instance is UNSATISFIABLE" << eom;
+      }
+    }
+  }
+
+  status=statust::UNSAT;
+  return resultt::P_UNSATISFIABLE;
+}
+
+void satcheck_ipasirt::set_assignment(literalt a, bool value)
+{
+  INVARIANT(!a.is_constant(), "cannot set an assignment for a constant");
+  INVARIANT(false, "method not supported");
+}
+
+satcheck_ipasirt::satcheck_ipasirt()
+: solver(nullptr)
+{
+  INVARIANT(!solver, "there cannot be a solver already");
+  solver=ipasir_init();
+}
+
+satcheck_ipasirt::~satcheck_ipasirt()
+{
+  if(solver)
+    ipasir_release(solver);
+  solver=nullptr;
+}
+
+bool satcheck_ipasirt::is_in_conflict(literalt a) const
+{
+  int v=a.var_no();
+
+  if(ipasir_failed(solver, v))
+      return true;
+  return false;
+}
+
+void satcheck_ipasirt::set_assumptions(const bvt &bv)
+{
+  forall_literals(it, bv)
+    if(it->is_true())
+    {
+      assumptions.clear();
```

Comment:
It seems assumptions can be cleared several times, I suggest using `std::find_if` instead.

Reply:
I'll replace it with a find_if.
However, the next line is a return, so that clear() is only called once.